### PR TITLE
ARROW-8766: [Python] Allow implementing filesystems in Python

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
A first test case is working, using fsspec's in-memory filesystem:

```
In [1]: import pyarrow.parquet as pq
   ...: import pyarrow.dataset as ds
   ...:
   ...: from pyarrow.fs import PyFileSystem
   ...: from pyarrow.tests.test_fs import FSSpecHandler

In [2]: import fsspec

In [3]: memfs = fsspec.filesystem("memory")

In [4]: table = pa.table({'a': [1, 2, 3]})

In [5]: with memfs.open("test", "wb") as f:
   ...:     pq.write_table(table, f)
   ...:

In [6]: dataset = ds.dataset("test", filesystem=PyFileSystem(FSSpecHandler(memfs)))

In [7]: dataset.to_table(filter=ds.field('a') > 1).to_pandas()
Out[7]:
   a
0  2
1  3
```

I only have a bit trouble finding a robust way to get a file handle for an open file from the fsspec filesystem, but so that is not related to this PR ;) For the rest it was quite easy to get this working!

File: python/pyarrow/_fs.pyx [hunk 1]
```diff
@@ -55,24 +56,56 @@ def _normalize_path(FileSystem filesystem, path):
...
+    def __init__(self, path, FileType type=FileType.Unknown, *,
+                 mtime=None, size=None):
+        self.info.set_path(tobytes(path))
+        self.info.set_type(_unwrap_file_type(type))
+        if mtime is not None:
+            if isinstance(mtime, datetime):
+                self.info.set_mtime(PyDateTime_to_TimePoint(
+                    <PyDateTime_DateTime*> mtime))
+            else:
+                self.info.set_mtime(TimePoint_from_ns(mtime))
```
Comment (pitrou) [inline python/pyarrow/_fs.pyx:89]: Note to self: should instead add a separate `mtime_ns` argument.

**Comment by jorisvandenbossche** [comment]
The WIP attempt to create a Handler for fsspec is here (which the example I posted above is using): https://github.com/jorisvandenbossche/arrow/commit/14ac33deff719ce368aea6d43bc134c4bcda8d61. With that, reading files is working. Although there are some incosistencies in the file-like objects the `open()` method creates, so it will probably not work for all fsspec filesystems. 
(writing is still TODO).

File: python/pyarrow/_fs.pyx [hunk 2]
```diff
@@ -685,3 +756,234 @@ cdef class _MockFileSystem(FileSystem):
...
+    @abstractmethod
+    def delete_file(self, path):
+        """
+        Implement PyFileSystem.delete_file(...).
+        """
+
+    @abstractmethod
+    def move(self, src, dest):
+        """
+        Implement PyFileSystem.move(...).
```
Comment (jorisvandenbossche) [inline python/pyarrow/_fs.pyx:865]: I suppose methods like this are expected to just raise the appropriate FileNotFoundError in case `src` does not exist?

**Review by jorisvandenbossche — APPROVED** [review]
@pitrou looks all good to me. Went through the code, and tested it further with the fsspec handler, and all seems to be working nicely. 
Only the `open_append_stream`, I am not yet fully sure how to do this (if possible). The `PythonFile` class might not yet support an append (mode="a") writable mode?

**Comment by jorisvandenbossche** [comment]
> The PythonFile class might not yet support an append (mode="a") writable mode?

Recognizing append mode as a "write" mode seems to work:

```diff
--- a/python/pyarrow/io.pxi
+++ b/python/pyarrow/io.pxi
@@ -654,6 +654,8 @@ cdef class PythonFile(NativeFile):
             kind = 'w'
         elif inferred_mode.startswith('r'):
             kind = 'r'
+        elif inferred_mode.startswith('a'):
+            kind = 'w'
         else:
             raise ValueError('Invalid file mode: {0}'.format(mode))
```

(if that's a proper fix, can include that in my follow-up PR)

File: cpp/src/arrow/python/filesystem.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      open_input_file;
+  std::function<void(PyObject*, const std::string& path,
+                     std::shared_ptr<io::OutputStream>* out)>
+      open_output_stream;
+  std::function<void(PyObject*, const std::string& path,
+                     std::shared_ptr<io::OutputStream>* out)>
+      open_append_stream;
+};
+
+class ARROW_PYTHON_EXPORT PyFileSystem : public arrow::fs::FileSystem {
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/python/filesystem.h:69]: It's not public as a C++ API, indeed, but if there's some stuff there that needs explaining, feel free to point it out and I'll add some comments.

File: python/pyarrow/_fs.pyx (hunk 2 above)
Comment (pitrou) [reply python/pyarrow/_fs.pyx:865]: Right.

**Comment by pitrou** [comment]
> Recognizing append mode as a "write" mode seems to work:

You'll have to check that writing indeed appends at the end rather than e.g. truncating.

More generally, while I originally added the append-open method, I'm not sure it will ever be useful, so in the meantime if an implementation wants to raise `NotImplementedError`, it's fine to me.

**Comment by jorisvandenbossche** [comment]
> You'll have to check that writing indeed appends at the end rather than e.g. truncating.

At least the tests pass (I added a fsspec local filesytem handler to the list of filesytem fixtures), and from a small local test it seems to be appending. 

Now, if it's not actually used within arrow at the moment (eg for enabling reading/writing parquet files or datasets with those filesystems, we won't need it), then it is also not very important to get this working.

File: cpp/src/arrow/python/filesystem.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply cpp/src/arrow/python/filesystem.h:69]: Maybe just one-liner indicating this is only used to implement pyarrow.fs.PyFileSystem could be added then, but not that important

**Pull request closed without merging**